### ISSUE TITLE
[E86] PyTorch Parity Testing for All Layers and Architectures

### ISSUE THREAD

**Issue by dndungu** [zerfoo/zerfoo#572]
### Tasks

- [x] T86.0.1 Wire Conv2D golden Go test  Est: 30m  verifies: [UC-L01]  DONE 2026-04-10
- [x] T86.0.2 Wire FFN golden Go test  Est: 30m  verifies: [UC-L01]  DONE 2026-04-10
- [x] T86.0.3 Wire BatchNorm golden Go test  Est: 30m  verifies: [UC-L01]  DONE 2026-04-10
- [x] T86.0.4 Wire Dropout golden Go test  Est: 15m  verifies: [UC-L01]  DONE 2026-04-10
- [x] T86.0.5 Wire AdamW optimizer golden Go test  Est: 30m  verifies: [UC-L02]  DONE 2026-04-10
- [x] T86.0.6 Wire SGD optimizer golden Go test  Est: 15m  verifies: [UC-L02]  DONE 2026-04-10
- [x] T86.0.7 Wire SimpleRNN golden Go test  Est: 30m  verifies: [UC-L01]  DONE 2026-04-10
- [x] T86.0.8 Wire S4 SSM golden Go test  Est: 30m  verifies: [UC-L01]  DONE 2026-04-10
- [x] T86.0.9 Wire MambaBlock golden Go test
- [x] T86.0.10 Wire TransformerBlock structural Go test  Est: 45m  verifies: [UC-L01]  DONE 2026-04-10
- [x] T86.0.11 Run go vet + go test on all wired tests  Est: 15m  verifies: [infrastructure]  DONE 2026-04-10
- [x] T86.1.1 FastGelu activation  Est: 30m  verifies: [UC-L01]  DONE 2026-04-10
- [x] T86.1.2 SimplifiedLayerNorm  Est: 30m  verifies: [UC-L01]  DONE 2026-04-10
- [x] T86.1.3 SkipSimplifiedLayerNorm  Est: 30m  verifies: [UC-L01]  DONE 2026-04-10
- [x] T86.1.4 GQA (GroupedQueryAttention)
- [x] T86.1.5 AttentionHead structural  Est: 30m  verifies: [UC-L01]  DONE 2026-04-10
- [x] T86.1.6 MoE (MixtureOfExperts + MoEGate)
- [x] T86.1.7 LMHead  Est: 30m  verifies: [UC-L01]  DONE 2026-04-10
- [x] T86.1.8 MIMOMambaBlock
- [x] T86.1.9 AttnRes structural  Est: 30m  verifies: [UC-L01]  DONE 2026-04-10
- [x] T86.1.10 BlockAttnRes residual
- [x] T86.1.11 HModule hierarchical residual
- [x] T86.1.12 PatchEmbed timeseries  Est: 30m  verifies: [UC-L01]  DONE 2026-04-10
- [x] T86.1.13 GRN (Gated Residual Network)
- [x] T86.1.14 TSMixerBlock  Est: 30m  verifies: [UC-L01]  DONE 2026-04-10
- [x] T86.1.15 MLSTM structural  Est: 45m  verifies: [UC-L01]  DONE 2026-04-10
- [x] T86.1.16 SLSTM structural  Est: 45m  verifies: [UC-L01]  DONE 2026-04-10
- [x] T86.1.17 SSMLayer (timeseries)  Est: 30m  verifies: [UC-L01]  DONE 2026-04-10
- [x] T86.1.18 CLIPEncoder structural  Est: 1h  verifies: [UC-L01]  DONE 2026-04-10
- [x] T86.1.19 MelExtractor structural  Est: 30m  verifies: [UC-L01]  DONE 2026-04-10
- [x] T86.1.20 WhisperEncoder structural  Est: 1h  verifies: [UC-L01]  DONE 2026-04-10
- [x] T86.1.21 Core arithmetic ops (Add, Sub, Mul, Div, Pow, Sqrt, Sin, Cos)  Est: 45m  verifies: [UC-L01]  DONE...
- [x] T86.1.22 Core shape ops (Reshape, Concat)  Est: 45m  verifies: [UC-L01]  DONE 2026-04-10
- [ ] T86.1.23 Core comparison ops
- [x] T86.1.24 Run go vet + go test on all E86.1 tests  Est: 15m  verifies: [infrastructure]  DONE 2026-04-10
- [x] T86.2.1 Activation backward: ReLU, GELU, Sigmoid, Tanh, LeakyReLU, SwiGLU  Est: 1h  verifies: [UC-L01]  D...
- [x] T86.2.2 Normalization backward: LayerNorm, RMSNorm  Est: 45m  verifies: [UC-L01]  DONE 2026-04-10
- [x] T86.2.3 Core backward: Linear, MatMul  Est: 1h  verifies: [UC-L01]  DONE 2026-04-10
- [x] T86.2.4 Loss backward: MSE, BCE, CrossEntropy  Est: 45m  verifies: [UC-L01]  DONE 2026-04-10
- [x] T86.2.5 SSM backward: S4  Est: 1h  verifies: [UC-L01]  DONE 2026-04-10
- [x] T86.2.6 Attention backward: SDPA  Est: 45m  verifies: [UC-L01]  DONE 2026-04-10
- [x] T86.2.7 Run go vet + go test for all backward parity tests  Est: 15m  verifies: [infrastructure]  DONE 20...
- [x] T86.3.1 EMA: add golden + Go test for one EMA update  Est: 30m  verifies: [UC-L02]  DONE 2026-04-10
- [x] T86.3.2 SWA: add golden + Go test for one SWA step  Est: 30m  verifies: [UC-L02]  DONE 2026-04-10
- [x] T86.3.3 Initializers: statistical tests for Xavier, He  Est: 30m  verifies: [UC-L02]  DONE 2026-04-10
- [x] T86.3.4 Run go vet + go test for optimizer parity tests  Est: 15m  verifies: [infrastructure]  DONE 2026-...
- [x] T86.4.1 PatchTST structural  Est: 1h  verifies: [UC-L03]  DONE 2026-04-10
- [x] T86.4.2 N-BEATS structural  Est: 1h  verifies: [UC-L03]  DONE 2026-04-10
- [x] T86.4.3 DLinear golden-file parity  Est: 30m  verifies: [UC-L03]  DONE 2026-04-10
- [x] T86.4.4 ITransformer structural  Est: 1h  verifies: [UC-L03]  DONE 2026-04-10
- [x] T86.4.5 TFT structural  Est: 1h  verifies: [UC-L03]  DONE 2026-04-10
- [x] T86.4.6 CfC structural  Est: 1h  verifies: [UC-L03]  DONE 2026-04-10
- [x] T86.4.7 FTTransformer structural  Est: 1h  verifies: [UC-L03]  DONE 2026-04-10
- [x] T86.4.8 TabNet structural  Est: 1h  verifies: [UC-L03]  DONE 2026-04-10
- [x] T86.4.9 PPO structural  Est: 1h  verifies: [UC-L03]  DONE 2026-04-10
- [x] T86.4.10 SAC structural  Est: 1h  verifies: [UC-L03]  DONE 2026-04-10
- [x] T86.4.11 GCN structural  Est: 45m  verifies: [UC-L03]  DONE 2026-04-10
- [x] T86.4.12 GAT structural  Est: 45m  verifies: [UC-L03]  DONE 2026-04-10
- [x] T86.4.13 MarketVAE structural  Est: 1h  verifies: [UC-L03]  DONE 2026-04-10
- [x] T86.4.14 Run go vet + go test  Est: 15m  verifies: [infrastructure]  DONE 2026-04-10
- [x] T86.5.1 Build arm64 parity test image for DGX  Est: 30m  verifies: [infrastructure]  DONE 2026-04-11
- [x] T86.5.2 GPU vs CPU parity: activations  Est: 30m  verifies: [UC-L01]  DONE 2026-04-11
- [x] T86.5.3 GPU vs CPU parity: normalization  Est: 30m  verifies: [UC-L01]  DONE 2026-04-11
- [x] T86.5.4 GPU vs CPU parity: core ops  Est: 30m  verifies: [UC-L01]  DONE 2026-04-11
- [x] T86.5.5 GPU vs CPU parity: attention  Est: 45m  verifies: [UC-L01]  DONE 2026-04-11
- [x] T86.5.6 GPU vs CPU parity: RotaryEmbedding  Est: 30m  verifies: [UC-L01]  DONE 2026-04-11
- [x] T86.5.7 GPU backward parity: all trained layers  Est: 1h  verifies: [UC-L01]  DONE 2026-04-11
- [ ] T86.5.8 Submit tests to DGX via Spark and collect results  Est: 30m  verifies: [infrastructure]
- [x] T86.6.1 Add CPU parity tests to CI workflow  Est: 30m  verifies: [infrastructure]  DONE 2026-04-10
- [x] T86.6.2 Add golden file staleness check to CI  Est: 30m  verifies: [infrastructure]  DONE 2026-04-10
- [x] T86.6.3 Create parity coverage report  Est: 30m  verifies: [infrastructure]  DONE 2026-04-10
- [x] T86.6.4 Run go vet + golangci-lint  Est: 15m  verifies: [infrastructure]  DONE 2026-04-10

---
Synced from [plan.md](https://github.com/zerfoo/zerfoo/blob/main/docs/plan.md).
